tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+use crate::key_source::KeySource;
+use crate::schema::decoded::{Decoded, Hex};
+use crate::schema::{
+    Hashes, Role, RoleType, Root, Signature, Signed, Snapshot, SnapshotMeta, Target, Targets,
+    Timestamp, TimestampMeta,
+};
+use crate::sign::Sign;
+use crate::transport::Transport;
+use crate::Repository;
+use chrono::{DateTime, Utc};
+use olpc_cjson::CanonicalFormatter;
+use ring::digest::{digest, SHA256, SHA256_OUTPUT_LEN};
+use ring::rand::{SecureRandom, SystemRandom};
+use serde::Serialize;
+use serde_json::Value;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::HashMap;
+use std::num::NonZeroU64;
+use std::os::unix::fs::symlink;
+use std::path::Path;
+use walkdir::WalkDir;
+
+const SPEC_VERSION: &str = "1.0.0";
+
+// A map of key ID (from root.json) to its corresponding signing key
+type RootKeys = HashMap<Decoded<Hex>, Box<dyn Sign>>;
+
+/// A signed role, including its serialized form (`buffer`) which is meant to
+/// be written to file. The `sha256` and `length` are calculated from this
+/// buffer and included in metadata for other roles, which makes it
+/// imperative that this buffer is what is written to disk.
+///
+/// Convenience methods are provided on `SignedRepository` to ensure that
+/// each role's buffer is written correctly.
+#[derive(Debug, Clone)]
+pub struct SignedRole<T> {
+    signed: Signed<T>,
+    buffer: Vec<u8>,
+    sha256: [u8; SHA256_OUTPUT_LEN],
+    length: u64,
+}
+
+impl<T> SignedRole<T>
+where
+    T: Role,
+{
+    pub(crate) fn write<P>(&self, outdir: P, consistent_snapshot: bool) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let outdir = outdir.as_ref();
+        std::fs::create_dir_all(&outdir).context(error::DirCreate { path: &outdir })?;
+
+        let filename = match T::TYPE {
+            RoleType::Targets => {
+                if consistent_snapshot {
+                    format!("{}.targets.json", self.signed.signed.version())
+                } else {
+                    "targets.json".to_string()
+                }
+            }
+            RoleType::Snapshot => {
+                if consistent_snapshot {
+                    format!("{}.snapshot.json", self.signed.signed.version())
+                } else {
+                    "targets.json".to_string()
+                }
+            }
+            RoleType::Timestamp => "timestamp.json".to_string(),
+            RoleType::Root => format!("{}.root.json", self.signed.signed.version()),
+        };
+
+        let path = outdir.join(filename);
+        std::fs::write(&path, &self.buffer).context(error::FileWrite { path })
+    }
+}
+
+/// A set of signed TUF Repository metadata.
+///
+/// This metadata represents a signed TUF repository, but without the actual
+/// targets, the repository cannot be used. It is up to ther user to ensure
+/// all the target files are in the proper directories along with the
+/// contained metadata. There are convenience methods provided to help with
+/// this.
```

Comment:
```suggestion
/// This metadata represents a signed TUF repository and provides the ability to write the metadata to disk.
///
/// Note: without the target files, the repository cannot be used. It is up to the user to ensure
/// all target files referenced by the metadata are available. There are convenience methods to help with this.
```
Slight simplification, and not presenting a "negative" first.